doc/guides/modules.rst

```diff
@@ -10,14 +10,342 @@ system those are called *modules*. These modules must be integrated with the
+Zephyr module repository. In particular, a module code owner will
+
+* coordinate code reviewing in the module repository
+* be the default assignee in pull-requests against the repository's
+  main branch
+* request additional collaborators to be added to the repository, as
+  they see fit
+* regularly synchronize the module repository with its upstream
+  counterpart following the policies described in
+  :ref:`modules_synchronization`
+* be aware of security vulnerability issues in the external project
+  and update the module repository to include security fixes, as
+  soon as the fixes are available in the upstream code base
+* list any known security vulnerability issues, present in the
+  module code base, in Zephyr release notes.
+
+
+  .. note::
+
+     Module code owners are not required to be Zephyr
+     :ref:`Maintainers <project_roles>`.
+
+**Merging:** The Zephyr Release Engineering team has the right and the
+responsibility to merge pull requests in the main branch of a module
+repository.
+
+  .. note::
+
+     The merging of pull requests in module repositories shall be coupled
+     with the corresponding manifest file updates in the Zephyr main tree.
+
+
+Maintaining module code base
+============================
+
+Updates in the Zephyr main tree, for example, in public Zephyr APIs,
+may require patching a module code base. The responsibility for keeping
+the module code base up to date is shared between the **contributor** of
+such updates in Zephyr and the module **code owner**. In particular:
+
+* the contributor of the original changes in Zephyr is required to submit
+  the corresponding patches that are required in module repositories, to
+  ensure that Zephyr CI on the pull request with the original changes, as
+  well as the module integration testing are successful.
+
+* the module code owner has the overall responsibility for maintaining
+  the module code base in-synch with the Zephyr main tree, by doing, for
+  example, occassionally, more advance testing of the module code base.
+  The code onwer is required to fix issues in the module code base that
+  have not been caught by Zephyr pull request CI runs.
+
+  .. note::
+
+    The above polices apply on all Zephyr modules. However, modules which
+    Zephyr depends on, for example vendor HALs, are more crucial to be kept
+    up-to-date with Zephyr main tree.
+
+
+
+Contributing changes to modules
+===============================
+
+Changes submitted to a module's code base should be limited to
+
+* patches required due to updates in the Zephyr main tree
+* urgent patches, for example those that fix security vulnerability issues,
+  and, therefore, should not wait to be fixed, first, in the external project.
+
+Non trivial changes to module code base, including changes in the module
+design or functionality should be discouraged, if the module has an upstream
+project repository. In that case, such changes shall be submitted to the
+upstream project, directly.
```

Comment:
Similar comment about commas here. Mixing "discouraged" and "shall" is also contradictory.

Something like this may be clearer:

```suggestion
Non-trivial changes to a module's code base are discouraged.
This includes changes to the module's design or functionality.
Such changes should be submitted to the upstream project directly.
```

I don't think it's necessary to say "if the module has an upstream project repository" because it's already established above that zephyr-only code shall not live in a separate module repo.

I'll stop here, but I suggest going through the changes and trying to phrase them as simpler sentences without as many commas. This may help readability for both native and non-native speakers :smile: .

Reply:
I am not sure whether **discouraged** is ok, we should probably now allow that.

We already have experience with tinycbor, where we have used mynewt version as the module, in mynewt it has been part of codebase (a subdirectory), because we needed compatibility with mcumgr.
Because intel/tinycbor has been reluctant to accept changes required by mynewt, mynewt has basically added the features, diverging from intel/tinycbor to such a point that it has been really hard to switch back. 

At the same time some additional features have been implemented by us and finally we end up with fork that has been nightmarish to update from upstream.

Now we have intel/tinycbor fork that in reality has been overwritten with the code from mynewt.

Reply:
No, we do not want to do that, this was discussed in the meeting.
The reason is that if we accept that the downstream module has functionality that is very different from the upstream, essentially, this is not a zephyr module any more, and the code should be in the main tree.